### PR TITLE
Use GDisk to clean MBR/GPT before writing

### PR DESCRIPTION
**Pull request by kendase3**
Addresses OVER-5210

### CODE REVIEW

**Comment by kendase3** [comment]
This won't work until I've uploaded the gdisk goodies, but WoMM.  Feel free to take a gander.

File: src/gondar.cc [hunk 1]
```diff
@@ -2280,13 +2284,33 @@ DeviceGuyList GetDeviceList() {
   return device_list;
 }
 
+bool clearMbrGpt(char * physical_path) {
+  std::string physical_path_str(physical_path);
+  GPTData gptdata(physical_path_str);
+  // let's get some before/after data on the state of the gpt
+  printf("problems with gpt before reformat=%d\n", gptdata.Verify());
```
Comment (nicholasbishop) [inline src/gondar.cc]: `printf` -> `LOG_*`

File: src/gondar.cc [hunk 2]
```diff
@@ -2280,13 +2284,33 @@ DeviceGuyList GetDeviceList() {
...
+  gptdata.ClearGPTData();
+  gptdata.MakeProtectiveMBR();
+  int quiet = true;
+  gptdata.SaveGPTData(quiet);
+  printf("problems with gpt after reformat=%d\n", gptdata.Verify());
+  return true;
```
Comment (nicholasbishop) [inline src/gondar.cc]: Looks like `true` is the only possible return value?

**Comment by josephsussman** [comment]
Feel free to take a gondar

**Comment by kendase3** [comment]
PAL = Partition Alignment Linter

**Comment by kendase3** [comment]
The only outstanding issue I see at the moment is that gdisk has a variable shadowed and we don't tolerate such things.

File: infra/gdisk.cmake [hunk 3]
```diff
@@ -0,0 +1,29 @@
+
+cmake_minimum_required(VERSION 3.4)
+project(gdisk)
```
Comment (nicholasbishop) [inline infra/gdisk.cmake]: These two lines aren't needed for an included file

File: infra/gdisk.cmake [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+#include_directories(${CMAKE_CURRENT_SOURCE_DIR}/gdisk)
+#file(GLOB gdisk_sources "${CMAKE_CURRENT_SOURCE_DIR}/gdisk/*.cc")
+#add_library(gdisk STATIC ${gdisk_sources})
```
Comment (nicholasbishop) [inline infra/gdisk.cmake]: Commented-out code

**Comment by kendase3** [comment]
see https://github.com/neverware/gdisk/pull/3

File: infra/gdisk.cmake [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+  gdisk/support.cc
+  gdisk/guid.cc
+  gdisk/gpt.cc
+  gdisk/gptpart.cc
+  gdisk/mbrpart.cc
+  gdisk/mbr.cc
+  gdisk/crc32.cc
+  gdisk/parttypes.cc
+  gdisk/attributes.cc
+  gdisk/diskio.cc)
```
Comment (nicholasbishop) [inline infra/gdisk.cmake]: You can disable specific warnings for this target as needed. Something like:

```
target_compile_options(gdisk PRIVATE -Wno-shadow)
```

Comment (kendase3) [inline infra/gdisk.cmake]: we can do it that way.

File: src/gondar.cc [hunk 6]
```diff
@@ -2286,7 +2288,15 @@ bool Install(DeviceGuy* target_device,
   uint64_t device_num = target_device->device_num;
   uint64_t sector_size = GetSectorSize(device_num);
   uint64_t drive_size = GetDriveSize(device_num);
+  // FIXME: this is a leak
   char* physical_path = GetPhysicalName(device_num);
+  printf("using physical_path=%s\n", physical_path);
+  if (!clearMbrGpt(physical_path)) {
+    LOG_WARNING << "error clearing mbr/gpt" << std::endl;
```
Comment (nicholasbishop) [inline src/gondar.cc]: Explicit `endl`s aren't needed, the logger puts newlines in automatically.

File: src/gpt_pal.cc [hunk 7]
```diff
@@ -0,0 +1,25 @@
+
```
Comment (nicholasbishop) [inline src/gpt_pal.cc]: Copyright header

**Comment by nicholasbishop** [comment]
Missing update to https://github.com/neverware/gondar/blob/master/.gitmodules

File: infra/gdisk.cmake [hunk 8]
```diff
@@ -0,0 +1,22 @@
+
```
Comment (nicholasbishop) [inline infra/gdisk.cmake]: copyright header

File: docker/gondar-win32.Dockerfile [hunk 9]
```diff
@@ -6,6 +6,7 @@ ADD CMakeLists.txt Makefile /opt/gondar/
 ADD infra /opt/gondar/infra
 ADD minizip /opt/gondar/minizip
 ADD plog /opt/gondar/plog
+ADD gdisk /opt/gondar/gdisk
```
Comment (nicholasbishop) [inline docker/gondar-win32.Dockerfile]: Alphabetical order pls

Comment (nicholasbishop) [inline docker/gondar-win32.Dockerfile]: This `ADD` line is also needed in `gondar-linux.Dockerfile`

File: CMakeLists.txt [hunk 10]
```diff
@@ -129,7 +131,7 @@ target_compile_options(app PRIVATE ${EXTRA_WARNINGS})
 target_include_directories(app SYSTEM PUBLIC minizip plog/include)
 target_include_directories(app PRIVATE ${CMAKE_BINARY_DIR}/src)
 target_link_libraries(app PUBLIC
-  Qt5::Network Qt5::Widgets minizip minizip_extra)
+  Qt5::Network Qt5::Widgets minizip minizip_extra gdisk)
```
Comment (nicholasbishop) [inline CMakeLists.txt]: The CMake wrapper for gdisk doesn't build on Linux, can we move this library dependency down to the if-windows block below?

**Comment by nicholasbishop** [comment]
make format pls

File: infra/gdisk.cmake [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+  gdisk/crc32.cc
+  gdisk/parttypes.cc
+  gdisk/attributes.cc
+  gdisk/diskio.cc
+  src/gpt_pal.cc)
+
+target_compile_options(gdisk PRIVATE
+  -D_FILE_OFFSET_BITS=64)
+
+target_compile_options(gdisk PRIVATE -Wno-shadow)
```
Comment (nicholasbishop) [inline infra/gdisk.cmake]: This is no longer needed right?

File: src/gondar.cc [hunk 12]
```diff
@@ -2286,7 +2288,15 @@ bool Install(DeviceGuy* target_device,
   uint64_t device_num = target_device->device_num;
   uint64_t sector_size = GetSectorSize(device_num);
   uint64_t drive_size = GetDriveSize(device_num);
+  // FIXME: this is a leak
   char* physical_path = GetPhysicalName(device_num);
+  printf("using physical_path=%s\n", physical_path);
```
Comment (nicholasbishop) [inline src/gondar.cc]: printf -> LOG

File: infra/gdisk.cmake [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+  gdisk/support.cc
+  gdisk/guid.cc
+  gdisk/gpt.cc
+  gdisk/gptpart.cc
+  gdisk/mbrpart.cc
+  gdisk/mbr.cc
+  gdisk/crc32.cc
+  gdisk/parttypes.cc
+  gdisk/attributes.cc
+  gdisk/diskio.cc
```
Comment (nicholasbishop) [inline infra/gdisk.cmake]: Alphabetize pls

File: src/gpt_pal.cc [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+#include "gpt_pal.h"
+
+// We use gdisk to clean up the GPT such that Windows is happy writing to
+// the disk
+#include "../gdisk/gpt.h"
+
+bool clearMbrGpt(char* physical_path) {
+  std::string physical_path_str(physical_path);
+  GPTData gptdata(physical_path_str);
+  // let's get some before/after data on the state of the gpt
```
Comment (nicholasbishop) [inline src/gpt_pal.cc]: Is this comment out of date?

File: src/gpt_pal.h [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+
+#ifndef GPT_PAL_H
+#define GPT_PAL_H
+
+// We use gdisk to clean up the GPT such that Windows is happy writing to
+// the disk
+
+// This function is kept in a separate file to avoid an issue with pragmas
+// within gdisk affecting rufus-based functionality (getting disk extents)
+bool clearMbrGpt(char* physical_path);
```
Comment (nicholasbishop) [inline src/gpt_pal.h]: `const char`

**Pull request merged**